_posts/2022-05-14-appsync-graphql-subscriptions.md

```diff
@@ -0,0 +1,308 @@
+
+## Troubleshooting
+
+1- The subscription response must be optional. AppSync allows us to successfully define the schema with a required subscription response, but when we hook the the subscription to our frontend client, it throw an error that subscription responses must be optional.
+
+This will result to failure and error in subscription:
+
+```
+type Subscription {
+    getOrderStatus($orderId: ID!): Order! @aws_subscribe(mutations: ["updateOrderStatus"])
+}
+```
+
+2- A subscription response contains only the predicted fields which are returned by the mutation, all the other fields will be null. For instance, if we have a mutation which returns the `orderStatus` only and not the entire order object:
+
+```
+type Mutation {
+  updateOrderStatus(orderId: ID!, orderStatus: OrderStatus!): OrderStatus!
+}
+```
+
+The following subscription will send `null` for all requested data except `OrderStatus`.
+
+```
+export const GET_ORDER_STATUS = gql`
+  subscription GetOrderStatus($orderId: ID) {
+    getOrderStatus(orderId: $orderId) {
+      id
+      orderItems {
+        item
+        quantity
+      }
+      OrderStatus
+      createdAt
+    }
+  }
+`;
+```
+
+data:
+
+```
+result:{
+    data:{
+      getOrderStatus:{
+        OrderStatus: 'IN_PROGRESS',
+        orderItems: null, // orderItems is null because it was not requested in the mutation
+        createdAt: null   // createdAt is null because it was not requested in the mutation
+      }
+    }
+  }
+
+```
+
+3- Subscription arguments are indeed filters that specify when the server should notify the client.
+
+The way we have defined our subscription means that the client is interested in updates on order status when a specific orderId argument is passed through.
+
+However, if you want to enable the client to subscribe to a single order `OR` all orders, you can make this argument optional by removing the exclamation point (!) from the subscription prototype.
+
+```
+type Subscription {
+    getOrderStatus($orderId: ID): Order @aws_subscribe(mutations: ["updateOrderStatus"])
+}
+```
+
+Additionally, if you want clients to explicitly subscribe to all status updates for all orders, you should remove the argument as follows:
+
+```
+type Subscription {
+    getOrderStatus: Order
+    @aws_subscribe(mutations: ["updateOrderStatus"])
+}
+```
+
+Simply,if you want to know about all orders that are created, you could do the following:
+
+```
+type Subscription {
+    getNewOrder: Order
+        @aws_subscribe(mutations: ["addOrder"])
+}
+```
+
+The golden point here to consider is that, if the filtering field is not specified in the mutation response, the subscription is not fired! If we define the addOrder mutation as below ( createdAt filed has intentionally been removed from mutation response):
```

Comment:
The golden point here to consider is that,
should be
The key point to consider here is that,